### PR TITLE
feat: allow interpolation of cluster generator values

### PR DESCRIPTION
**Pull request by blakepettersson**
Allow the interpolation of `values` found in the cluster generator.
This allows interpolation of `{{name}}`, `{{server}}`,
`{{metadata.labels.*}}` and `{{metadata.annotations.*}}`. See
argoproj/applicationset#371.

This interpolation could potentially be extended to the list and
duck-type generators if desired.

Note on DCO:

If the DCO action in the integration test fails, one or more of your commits are not signed off. Please click on the *Details* link next to the DCO action for instructions on how to resolve this.

Checklist:

* [ ] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [ ] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [ ] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [x] Does this PR require documentation updates?
* [x] I've updated documentation as required by this PR.
* [ ] Optional. My organization is added to USERS.md.
* [x] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/tree/master/community#contributing-to-argo)
* [x] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [x] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).

### CODE REVIEW

**Review by crenshaw-dev — CHANGES_REQUESTED** [review]
This is super interesting, thanks for writing it up! Could you add some docs about how it would be used?

**Comment by blakepettersson** [comment]
> This is super interesting, thanks for writing it up! Could you add some docs about how it would be used?

Thanks for finding it useful 😄 I wrote some basic usage docs in `Generators-Clusters.md`, let me know if that is enough or if I need to complement with something else!

**Comment by blakepettersson** [comment]
@crenshaw-dev is there anything else I should do on this PR?

**Comment by crenshaw-dev** [comment]
@blakepettersson ah! No, besides pinging me to remind me to review. Thanks. :-)

File: applicationset/generators/cluster.go [hunk 1]
```diff
@@ -139,6 +151,16 @@ func (g *ClusterGenerator) GenerateParams(
 	return res, nil
 }
 
+func replaceTemplatedString(value string, params map[string]string, key string) error {
+	fstTmpl := fasttemplate.New(value, "{{", "}}")
+	replacedTmplStr, err := render.Replace(fstTmpl, params, true)
+	if err != nil {
+		return err
+	}
+	params[fmt.Sprintf("values.%s", key)] = replacedTmplStr
```
Comment (crenshaw-dev) [inline applicationset/generators/cluster.go:160]: I'm a little worried about setting directly to `params`. At this point, AppSet is admin-only, but in the future I'd like it to see it made available to lower-privileged users. In that case, I'd be concerned about a billion-laughs attack.

The attack would work like this:

```yaml
values:
  lol1: lol
  lol2: '{{values.lol1}}{{values.lol1}}{{values.lol1}}{{values.lol1}}{{values.lol1}}{{values.lol1}}{{values.lol1}}{{values.lol1}}{{values.lol1}}{{values.lol1}}'
  lol3: '{{values.lol2}}{{values.lol2}}{{values.lol2}}{{values.lol2}}{{values.lol2}}{{values.lol2}}{{values.lol2}}{{values.lol2}}{{values.lol2}}{{values.lol2}}'
  # ...etc. through lol9
```

There are probably other/easier ways to attack the AppSet controller, but I'd rather not add another one.

What if we just keep track of an `interpolatedParams` map which isn't merged back with `params` until all the other interpolation is finished?

**Review by crenshaw-dev — CHANGES_REQUESTED** [review]

File: applicationset/generators/cluster.go (hunk 1 above)
Comment (blakepettersson) [reply applicationset/generators/cluster.go]: > What if we just keep track of an interpolatedParams map which isn't merged back with params until all the other interpolation is finished?

Ah, good catch! I updated the PR to do just that, and added a test case for it as well.

**Review by crenshaw-dev — APPROVED** [review]
lgtm, thanks for your patience on the review!

**Pull request merged**